### PR TITLE
Workaround for pydantic v2

### PR DESCRIPTION
**Pull request by xl0**
It's a bit heavier than expected to make mypy happy on both pydantic versions.

### CODE REVIEW

File: pyairtable/models/_base.py
Comment (mesozoic) [inline pyairtable/models/_base.py:99]: I was able to drop the `unused-ignore` portion of this without reintroducing any failures.

File: tests/test_api_types.py
Comment (mesozoic) [inline tests/test_api_types.py:6]: Should this be `import pydantic.v1 as pydantic`, since mypy is going to run in a venv with `pydantic>=2`?

I'm not sure it winds up making a difference because we now have `# type: ignore` everywhere, but it seems "more correct", and perhaps it will make a difference at some point in the future.

File: pyairtable/api/types.py
Comment (mesozoic) [inline pyairtable/api/types.py:15]: Should we move this somewhere (maybe `pyairtable.utils` or a new module `pyairtable.compat`) so it only happens once, and then import from there in the rest of our code?

**Review by mesozoic — CHANGES_REQUESTED** [review]
Fantastic! 🙌  Thank you for figuring this out! 

I like this approach because it is agnostic which version our library users require for other parts of their project. Also, since it does not require changing any of our APIs or behaviors, we can make it a patch release instead of a minor version bump.

I suggested a couple changes below. I also think we could add a few lines to `tox.ini` to ensure that our test suite runs successfully against both major versions of pydantic:

```diff
diff --git a/tox.ini b/tox.ini
index 584ae4d..142db2e 100644
--- a/tox.ini
+++ b/tox.ini
@@ -3,6 +3,7 @@ envlist =
     pre-commit
     mypy
     py3{8,9,10,11}-requests{min,max}
+    py3{8,9,10,11}-pydantic1
     coverage
 
 [gh-actions]
@@ -29,6 +30,7 @@ deps =
     -r requirements-test.txt
     requestsmin: requests==2.22.0  # Keep in sync with setup.cfg
     requestsmax: requests>=2.22.0  # Keep in sync with setup.cfg
+    pydantic1: pydantic<2  # Lots of projects still use 1.x
 
 [testenv:coverage]
 passenv = COVERAGE_FORMAT
```

I'll be very happy to merge this branch once we've got test coverage and have at least tried those other ideas.

File: pyairtable/models/_base.py
Comment (xl0) [reply pyairtable/models/_base.py:99]: Are you sure you tried with v2? I think v2's BaseModel does not have __fields__, so mypy complains here.

File: tests/test_api_types.py
Comment (xl0) [reply tests/test_api_types.py:6]: This would cause a mypy error for v1 users, would it not?

File: pyairtable/api/types.py
Comment (xl0) [reply pyairtable/api/types.py:15]: I think this is a natural reaction to seeing duplicated code, but I feel in this case, the abstraction will just make it harder to follow what's going on.

**Comment by xl0** [comment]
Agree re tox, I'll add it later today.

File: pyairtable/api/types.py
Comment (xl0) [reply pyairtable/api/types.py:15]: Actually, since you are adding more pedantic, it definitely makes sense to move this into one place. I will do it.

File: pyairtable/models/_base.py
Comment (mesozoic) [reply pyairtable/models/_base.py:99]: 🤷  I'm running `tox -e mypy` and it's using the latest pydantic version, but I'm not sure it's worth our time to keep tweaking this particular nuance. Having `# type: ignore` feels kinda icky regardless, so who cares which options it needs? 😁

File: tests/test_api_types.py
Comment (mesozoic) [reply tests/test_api_types.py:6]: Hmm, I was (selfishly) thinking that since we run `tox -e mypy` in a separate environment, it will always use the latest pydantic. But you're right, we also have to think about library users who are running mypy on their own projects (and might have pinned dependencies to pydantic v1 _or_ v2).

I'm uncertain of the best way to handle this, but we probably want to test this branch as a dependency within some live code (that uses mypy) before we merge it. I can try to get to that in a week or so, but if you can sooner, all the better.

File: pyairtable/api/types.py
Comment (mesozoic) [reply pyairtable/api/types.py:15]: Thanks! I think most library users are not going to peek under the hood at how we're using pydantic; they're just going to want to import `Base` and `Table` and get on with life. So if it makes it easier for us (and I think it will, given how many places I'm starting to use pydantic) it's worth the time.

File: pyairtable/models/_base.py
Comment (xl0) [reply pyairtable/models/_base.py:99]: Yeah, it needs ignore-unused for v1 because the warning only occurs with pydantic v2.

**Comment by xl0** [comment]
@mesozoic I updated the PR to
- Add -pydantic1 to tox tests. Also added mypy-pydantic1.
- Moves pydantic stuff into pyairtable/pydantic.py. Can't be in utils due to circular dependency.

**Comment by xl0** [comment]
@mesozoic , any idea what's up with the docs failing?

**Comment by mesozoic** [comment]
> @mesozoic , any idea what's up with the docs failing?

I think you might've forgotten to `git add` the new file. Can we call it `_compat.py`, though, to (1) make it clear that it's internal and not part of the public API, and (2) to avoid any confusion arising from the module and its contents having the same name?

**Comment by xl0** [comment]
@mesozoic , heh, my bad. Ok, renamed it to _compat.py.

File: tests/test_api_types.py
Comment (mesozoic) [reply tests/test_api_types.py]: I used this branch in a project of mine with both `pydantic~=1` and `pydantic~=2`, and the test suite (including mypy) passed in both cases. 🎉

**Review by mesozoic — APPROVED** [review]

**Comment by mesozoic** [comment]
@xl0 One more request, could you rebase the branch with [signed commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification)? Easiest way is with SSH, in my experience, but the repo requires it before we merge. I could rebase and merge it on your behalf, but then you won't get any credit. 😁

**Comment by xl0** [comment]
Ok, I'll check it later today.

**Comment by xl0** [comment]
@mesozoic , done

**Comment by mesozoic** [comment]
Thanks very much for solving this!

**Pull request merged**
